crates/neon-runtime/src/napi/fun.rs

```diff
@@ -1,29 +1,81 @@
 //! Facilities for working with JS functions.
 
+use std::mem::MaybeUninit;
 use std::os::raw::c_void;
-use std::ptr::null;
+use std::ptr;
 
-use crate::call::CCallback;
 use crate::napi::bindings as napi;
 use crate::raw::{Env, Local};
 
-/// Mutates the `out` argument provided to refer to a newly created `v8::Function`. Returns
-/// `false` if the value couldn't be created.
-pub unsafe fn new(out: &mut Local, env: Env, callback: CCallback) -> bool {
+pub unsafe fn new<F>(env: Env, name: &str, callback: F) -> Result<Local, napi::Status>
+where
+    F: Fn(Env, napi::CallbackInfo) -> Local + 'static,
+{
+    let mut out = MaybeUninit::uninit();
+    let data = Box::into_raw(Box::new(callback));
     let status = napi::create_function(
         env,
-        null(),
-        0,
-        Some(std::mem::transmute(callback.static_callback)),
-        callback.dynamic_callback,
-        out as *mut Local,
+        name.as_ptr().cast(),
+        name.len(),
+        Some(call_boxed::<F>),
+        data.cast(),
+        out.as_mut_ptr(),
     );
 
-    status == napi::Status::Ok
+    if status == napi::Status::PendingException {
+        Box::from_raw(data);
+
+        return Err(status);
+    }
+
+    assert_eq!(status, napi::Status::Ok);
+
+    let out = out.assume_init();
+
+    #[cfg(feature = "napi-5")]
+    {
+        unsafe extern "C" fn drop_function<F>(
+            _env: Env,
+            _finalize_data: *mut c_void,
+            finalize_hint: *mut c_void,
+        ) {
+            Box::from_raw(finalize_hint.cast::<F>());
+        }
+
+        let status = napi::add_finalizer(
+            env,
+            out,
+            ptr::null_mut(),
+            Some(drop_function::<F>),
+            data.cast(),
+            ptr::null_mut(),
+        );
+
+        assert_eq!(status, napi::Status::Ok);
+    }
+
+    Ok(out)
 }
 
-pub unsafe fn get_dynamic_callback(_env: Env, data: *mut c_void) -> *mut c_void {
-    data
+unsafe extern "C" fn call_boxed<F>(env: Env, info: napi::CallbackInfo) -> Local
```

Comment:
A brief comment explaining the purpose of this function, i.e., what's boxed, would probably be good.